stdlib/streams/src/main/ballerina/streams/snapshot.bal

```diff
@@ -0,0 +1,350 @@
+    string[] files = getSnapshotFiles(persistancePath);
+    string? latestFile = ();
+    if (files.length() > 0) {
+        latestFile = files[files.length() - 1];
+    }
+    return latestFile;
+}
+
+# Function to purge/delete old snapshot files from persistance path
+# + persistancePath - A `string` path of the directory where the states should get persisted.
+function purgeOldSnapshotFiles(string persistancePath) {
+    int revisionsToKeep = 5;
+    string[] files = getSnapshotFiles(persistancePath);
+    if (files.length() > revisionsToKeep) {
+        int i = 0;
+        while (i < (files.length() - revisionsToKeep)) {
+            string f = files[i];
+            internal:Path p = new(f);
+            if (p.exists()) {
+                error? e = p.delete();
+                if (e is error) {
+                    io:println("couldn't delete snapshot ", p.getPathValue(), ". ", e.detail());
+                }
+            }
+            i += 1;
+        }
+    }
+}
+
+# Function to restore a persisted snapshotable states from a file.
+function restoreStates() {
+    if (!stateLoaded) {
+        string? lastSnapshotFile = getLatestSnapshotFile(persistanceDirectory);
+        if (lastSnapshotFile is string) {
+            string stateStr = readFile(lastSnapshotFile, "UTF8");
+            var retrievedState = trap deserialize(stateStr);
+            if (retrievedState is map<map<any>>) {
+                streamsPersistanceState = retrievedState;
+                stateLoaded = true;
+            }
+        }
+    }
+    foreach var (k, v) in streamsPersistanceState {
+        boolean loaded = loadedStates[k] ?: false;
+        if (!loaded) {
+            Snapshotable? s = snapshotables[k];
+            if (s is Snapshotable) {
+                s.restoreState(v);
+                loadedStates[k] = true;
+            }
+        }
+    }
+}
+
+# Function to iterate through all the snapshotables, and persist their states into a file.
+# + return - An `error` if the state cannot be persisted into a file.
+function persistStatesAndPurgeOldSnapshots() returns error? {
+    foreach var (k, v) in snapshotables {
+        streamsPersistanceState[k] = v.saveState();
+    }
+    error? e = writeStateToFile(persistanceDirectory);
+    purgeOldSnapshotFiles(persistanceDirectory);
+    return e;
+}
+
+# Function to initilize a Timer task and start periodic snapshotting.
+function startPersisting() {
+    (function () returns error?) onTriggerFunction = persistStatesAndPurgeOldSnapshots;
+    function (error) onErrorFunction = persitanceError;
+    persistTimer = new task:Timer(onTriggerFunction, onErrorFunction, persistanceIntervalInMillis,
+        delay = persistanceIntervalInMillis);
+    persistTimer.start();
+}
+
+# Function to restore state of a given object.
+# + k - An unique `string` identifier for the snapshotable reference.
+# + o - The snapshotable reference to be restored.
+public function restoreState(string k, any o) {
```

Comment:
Can we improve the parameter names? Since these are public functions too?

Reply:
Fixed in c37733b